shell/platform/android/io/flutter/embedding/engine/android/FlutterView.java

```diff
@@ -103,6 +128,176 @@ private void init() {
+    sendLocalesToFlutter(newConfig);
+    sendUserSettingsToFlutter();
+  }
+
+  /**
+   * Invoked when this {@code FlutterView} changes size, including upon initial
+   * measure.
+   *
+   * The initial measure reports an {@code oldWidth} and {@code oldHeight} of zero.
+   *
+   * Flutter cares about the width and height of the view that displays it on the host
+   * platform. Therefore, when this method is invoked, the new width and height are
+   * communicated to Flutter as the "physical size" of the view that displays Flutter's
+   * UI.
+   */
+  @Override
+  protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight) {
+    // TODO(mattcarroll): hookup to viewport metrics.
+    super.onSizeChanged(width, height, oldWidth, oldHeight);
+  }
+
+  /**
+   * Invoked when Android's desired window insets change, i.e., padding.
+   *
+   * Flutter does not use a standard {@code View} hierarchy and therefore Flutter is
+   * unaware of these insets. Therefore, this method calculates the viewport metrics
+   * that Flutter should use and then sends those metrics to Flutter.
+   *
+   * This callback is not present in API < 20, which means lower API devices will see
+   * the wider than expected padding when the status and navigation bars are hidden.
+   */
+  @Override
+  public final WindowInsets onApplyWindowInsets(WindowInsets insets) {
+    // TODO(mattcarroll): hookup to Flutter metrics.
+    return insets;
+  }
+
+  /**
+   * Invoked when Android's desired window insets change, i.e., padding.
+   *
+   * {@code fitSystemWindows} is an earlier version of
+   * {@link #onApplyWindowInsets(WindowInsets)}. See that method for more details
+   * about how window insets relate to Flutter.
+   */
+  @Override
+  @SuppressWarnings("deprecation")
+  protected boolean fitSystemWindows(Rect insets) {
+    // TODO(mattcarroll): hookup to Flutter metrics.
+    return super.fitSystemWindows(insets);
+  }
+  //------- End: Process View configuration that Flutter cares about. --------
+
+  //-------- Start: Process UI I/O that Flutter cares about. -------
+  /**
+   * Creates an {@link InputConnection} to work with a {@link android.view.inputmethod.InputMethodManager}.
+   *
+   * Any {@code View} that can take focus or process text input must implement this
+   * method by returning a non-null {@code InputConnection}. Flutter may render one or
+   * many focusable and text-input widgets, therefore {@code FlutterView} must support
+   * an {@code InputConnection}.
+   *
+   * The {@code InputConnection} returned from this method comes from a
+   * {@link TextInputPlugin}, which is owned by this {@code FlutterView}. A
+   * {@link TextInputPlugin} exists to encapsulate the nuances of input communication,
+   * rather than spread that logic throughout this {@code FlutterView}.
+   */
+  @Override
+  public InputConnection onCreateInputConnection(EditorInfo outAttrs) {
+    if (!isAttachedToFlutterEngine()) {
```

Comment:
When would this happen?

Reply:
When the View is added to the hierarchy before an engine has been provided.

Reply:
Can we document why we think we need to cover this and what it actually means?  I might be bogged down by notions from the iOS embedding where it would be impossible to have a View/ViewController that has no Engine - in such a case, the View's ViewController would have created an engine automatically with the default parameters.  If that doesn't happen here, we should document it (and we should maybe be warning developers somehow that even though they have a FlutterView they don't have any Dart backing it).